### PR TITLE
add join across sources and support xlsx

### PR DESCRIPTION
**Pull request by ahuang11**
Initially I tried doing a custom join analysis (code at the bottom) but I felt like it was duplicating a lot of code, and I didn't know how to call SQLAgent to do the auto join.

I also considered using the built-in JoinedTable source, but this again was semi-tedious vs letting the LLM automagically figuring out, not only the proper columns to join on, but also how to transform the column values so that it joins properly.

So I went the route where I support joins **across sources**.

This is the part where I experimented quite a bit, but I'm not sure what's best.

Reference https://duckdb.org/2024/01/26/multi-database-support-in-duckdb.html (I wonder if we can drop IntakeDremio usage and simply use all duckdb https://www.dremio.com/blog/using-duckdb-with-your-dremio-data-lakehouse/?)

### Copy all the necessary tables' data over into a separate, **temporary**, in-memory duckdb source

This is the current implementation. The pros of this method is that it likely will work with IntakeDremio, but not entirely sure if it's efficient and whether all the table memory can fit in memory.

```python
import duckdb
import pandas as pd

# Load data into pandas DataFrames
df_medals = pd.read_csv("olympic_medals.csv")
df_hosts = pd.read_csv("olympic_hosts.csv")

# Create two separate in-memory DuckDB connections
memory_1 = duckdb.connect(":memory:")
memory_2 = duckdb.connect(":memory:")

# Create tables in each connection
memory_1.from_df(df_medals).to_table("olympic_medals")
memory_2.from_df(df_hosts).to_table("olympic_hosts")

# Create a third in-memory DuckDB connection
memory_3 = duckdb.connect(":memory:")

# Transfer the olympic_medals table from memory_1 to memory_3
medals_table = memory_1.execute("SELECT * FROM olympic_medals").fetchdf()
memory_3.from_df(medals_table).to_table("olympic_medals")

# Transfer the olympic_hosts table from memory_2 to memory_3
hosts_table = memory_2.execute("SELECT * FROM olympic_hosts").fetchdf()
memory_3.from_df(hosts_table).to_table("olympic_hosts")

# Perform the join query in memory_3
query = """
SELECT *
FROM olympic_medals
JOIN olympic_hosts
ON olympic_medals.Year = olympic_hosts.Year
"""

# Execute the query and fetch the result into a DataFrame
result = memory_3.execute(query).fetchdf()

# Display the result
print(result)
```

### Copy all the necessary tables' data over into a **central**, in-memory (literally, `memory["central_source"]`) duckdb source

Similar to above, but we persist `memory_3` as `memory["central_source"]` and we simply keep adding more tables to it on upload / init. Then on the sidebar, potentially allow users to delete unused tables/sources.

```python
import duckdb
import pandas as pd

# Load data into pandas DataFrames
df_medals = pd.read_csv("olympic_medals.csv")
df_hosts = pd.read_csv("olympic_hosts.csv")

# Create a single in-memory DuckDB connection
con = duckdb.connect(":memory:")

# Load DataFrames into the same connection as tables
con.from_df(df_medals).to_table("olympic_medals")
con.from_df(df_hosts).to_table("olympic_hosts")

# Perform the join query within the same connection
query = """
SELECT *
FROM olympic_medals AS om
JOIN olympic_hosts AS oh
ON om.slug_game = oh.game_slug
"""

# Execute the query and fetch the result into a DataFrame
result = con.execute(query).fetchdf()

# Display the result
print(result)
```

### Persist sources onto disk and create views of it:

This depends on saving data to disk; not sure if that's desired, but you can simply "ATTACH_DATABASE"

```python
import duckdb
import pandas as pd

# Load data into pandas DataFrames
df_medals = pd.read_csv("olympic_medals.csv")
df_hosts = pd.read_csv("olympic_hosts.csv")

# Create named in-memory DuckDB connections
memory_1 = duckdb.connect("memory_1")
memory_2 = duckdb.connect("memory_2")

# Create tables in each connection
memory_1.from_df(df_medals).to_table("olympic_medals")
memory_2.from_df(df_hosts).to_table("olympic_hosts")

# Create a third in-memory DuckDB connection
memory_3 = duckdb.connect("memory_3")

# Attach the first two connections as databases to memory_3 with unique names
memory_3.execute("ATTACH DATABASE 'memory_1' AS mem1")
memory_3.execute("ATTACH DATABASE 'memory_2' AS mem2")

# Perform the join query in memory_3, referencing the attached databases
query = """
SELECT *
FROM mem1.olympic_medals AS om
JOIN mem2.olympic_hosts AS oh
ON om.slug_game = oh.game_slug
"""

# Execute the query and fetch the result into a DataFrame
result = memory_3.execute(query).fetchdf()

# Display the result
print(result)

```

---

image

image

Figure showing data.csv joined on an uploaded data2.xlsx

```python
import pandas as pd

df = pd.DataFrame({
    'column1': [1, 2, 3],
    'column2': ['A', 'B', 'C']
})

df2 = pd.DataFrame({
    'column3': [4, 5, 6],
    'column2': ['A', 'B', 'C']
})

df.to_csv("data.csv")

df2.to_excel("data2.xlsx")
```

```python
import io
import asyncio

import lumen.ai as lmai
import panel as pn
import pandas as pd

from lumen.sources.duckdb import DuckDBSource

pn.extension(
    "tabulator", "codeeditor", "vega", inline=False, template="fast", notifications=True,
)

llm = lmai.llm.OpenAI(interceptor_path="test.db")

source = DuckDBSource(
    tables=["data.csv"],
    uri=":memory:",
)
lmai.memory["current_source"] = source

assistant = lmai.Assistant(
    llm=llm,
    agents=[
        lmai.agents.SourceAgent,
        lmai.agents.TableAgent,
        lmai.agents.TableListAgent,
        lmai.agents.SQLAgent,
        lmai.agents.PipelineAgent,
        lmai.agents.hvPlotAgent,
        lmai.agents.ChatAgent,
    ],
    suggestions=["Join hosts and medals"]
)
assistant.servable("Lumen.ai")
assistant.controls().servable(area="sidebar")
```

---

```

class Join(Analysis):
    """
    Provides high level data profiling.
    """

    async def __call__(self, pipeline):
        def add_table(event):
            if input_tabs.active == 0 and upload.value:
                if upload.filename.endswith("csv"):
                    df = pd.read_csv(io.BytesIO(upload.value), parse_dates=True)
                elif upload.filename.endswith((".parq", ".parquet")):
                    df = pd.read_parquet(io.BytesIO(upload.value))
                elif upload.filename.endswith(".xlsx"):
                    df = pd.read_excel(io.BytesIO(upload.value))
                duckdb_source._connection.from_df(df).to_table(name.value)
                duckdb_source.tables = [name.value]
                memory["current_source"] = duckdb_source
            name.value = ""
            upload.value = None
            upload.filename = ""
            src = memory["current_source"]
            tables[:] = [
                (t, pn.widgets.Tabulator(src.get(t), sizing_mode="stretch_width")) for t in src.get_tables()
            ]
            input_column.visible = False

        def add_name(event):
            if "/" in event.new:
                name.value = event.new.split("/")[-1].split(".")[0].replace("-", "_")
            elif not name.value and event.new:
                name.value = event.new.split(".")[0].replace("-", "_")
                name.visible = True

        def enable_add(event):
            add.visible = not bool(name.value)

        name = pn.widgets.TextInput(name="Name your table", visible=False)
        tables = memory["current_source"].get_tables()
        select = pn.widgets.Select(name="Select table", options=tables)
        upload = pn.widgets.FileInput(align="end", accept=".csv,.parquet,.parq")
        add = pn.widgets.Button(
            name="Add table",
            icon="table-plus",
            visible=False,
            button_type="success",
        )
        input_tabs = pn.Tabs(("Select", select), ("Upload", upload), sizing_mode="stretch_width")
        tables = pn.Tabs(sizing_mode="stretch_width")
        duckdb_source = DuckDBSource(uri=":memory:")
        input_column = pn.Column(
            input_tabs,
            name,
            add,
            sizing_mode="stretch_width",
        )
        menu = pn.Column(
            input_column,
            tables,
            sizing_mode="stretch_width",
        )
        upload.param.watch(add_name, "filename")
        upload.param.watch(enable_add, "value")
        add.on_click(add_table)

        assistant.interface.send(menu, respond=False, user="SourceAgent")
        while not add.clicks:
            await asyncio.sleep(0.05)

```

### CODE REVIEW

**Comment by philippjfr** [comment]
>simply use all duckdb https://www.dremio.com/blog/using-duckdb-with-your-dremio-data-lakehouse/?)

We absolutely should.

File: lumen/ai/views.py [hunk 1]
```diff
@@ -83,7 +84,20 @@ async def _render_component(self):
...
+                tables = yaml_spec["source"]["tables"]
+                old_source = memory["current_source"]
+                old_source.tables = tables
+                pipeline.param.update(
+                    source=old_source,
+                    table=yaml_spec["table"]
+                )
+                self.component = pipeline
+            else:
+                self.component = type(self.component).from_spec(yaml_spec)
```
Comment (ahuang11) [inline lumen/ai/views.py:100]: Semi-hacky, but I don't know how to approach.

I also tried `pipeline.source.create_sql_expr_source` for a very, very long time, but I could not figure out why the stored tables in `_connection` were not transferring over... so instead, I just use the old source and manually update the tables + table.

**Comment by ahuang11** [comment]
Not sure how to serialize the uploaded data.

File: lumen/ai/agents.py [hunk 2]
```diff
@@ -552,21 +503,49 @@ async def _create_valid_sql(self, messages, system, source, tables, errors=None)
...
+                    print(f"Could not add to catalog {e}")
+        else:
+            source = next(iter(sources.values()))[0]
+
         # check whether the SQL query is valid
         expr_name = output.expr_name
         sql_expr_source = source.create_sql_expr_source({expr_name: sql_query})
-        pipeline = Pipeline(source=sql_expr_source, table=expr_name)
+        print(sql_query)
+        breakpoint()
```
Comment (ahuang11) [inline lumen/ai/agents.py:532]:
```suggestion
```

File: lumen/ai/agents.py [hunk 3]
```diff
@@ -588,34 +566,55 @@ async def answer(self, messages: list | str):
...
         sql_prompt = render_template(
             "sql_query.jinja2",
             tables_sql_schemas=tables_sql_schemas,
             dialect=dialect,
         )
         system = await self._system_prompt_with_context(messages) + sql_prompt
-        sql_query = await self._create_valid_sql(messages, system, source, tables)
-        memory["current_sql"] = sql_query
+        if join_required:
+            messages[-1]["content"] = re.sub(r"//[^/]+//", "", messages[-1]["content"])
```
Comment (philippjfr) [inline lumen/ai/agents.py:615]:
```suggestion
            # Remove source prefixes message, e.g. //<source>//<table>
            messages[-1]["content"] = re.sub(r"//[^/]+//", "", messages[-1]["content"])
```

File: lumen/ai/agents.py [hunk 4]
```diff
@@ -972,7 +972,7 @@ async def _system_prompt_with_context(
         system_prompt += f"\nHere are the columns of the table: {columns}"
         return system_prompt
 
-    async def answer(self, messages: list | str):
+    async def answer(self, messages: list | str, agents=None):
```
Comment (philippjfr) [inline lumen/ai/agents.py:975]:
```suggestion
    async def answer(self, messages: list | str, agents: list[Agent] = None):
```

File: lumen/ai/agents.py [hunk 5]
```diff
@@ -552,21 +503,48 @@ async def _create_valid_sql(self, messages, system, source, tables, errors=None)
...
+                relation = source._connection.from_df(df)
+                try:
+                    # DREMIO.SOMETHING turns into DREMIO_SOMETHING
+                    renamed_table = a_table.replace(".", "_")
+                    sql_query = sql_query.replace(a_table, renamed_table)
+                    relation.to_table(renamed_table)
+                except duckdb.CatalogException as e:
+                    print(f"Could not add to catalog {e}")
+        else:
+            source = next(iter(sources.values()))[0]
```
Comment (philippjfr) [inline lumen/ai/agents.py:526]:
```suggestion
            source = next(iter(sources.values()))
```

File: lumen/ai/analysis.py [hunk 6]
```diff
@@ -12,11 +16,71 @@ class Analysis(param.ParameterizedFunction):
...
+        self._run_button = source_controls[0][-1]
+        name_input = source_controls[0][1]
+        name_input.link(self, value='table_name')
+
+        source = memory.get("current_source")
+        table = memory.get("current_table")
+        self._previous_source = source
+        self._previous_table = table
+        columns = list(get_schema(source, table=table).keys())
+        print(columns)
```
Comment (philippjfr) [inline lumen/ai/analysis.py:65]:
```suggestion
```

**Comment by philippjfr** [comment]
Okay, pushed a bunch of fixes and improvements here:

- `DuckDBSource.create_sql_expr_source` now has an option to materialize temporary tables so the LLM can refer to tables by name
- Refactored the SQLAgent to pull out a lot of the join logic and emphasize that it should only join when it really needs to.
- Skip search for correct analysis when the user clicks one of the buttons
- Allow assistant to select different source if table can't be found in `current_source`

**Pull request merged**